rama-tls-acme/src/client.rs

```diff
@@ -0,0 +1,681 @@
+
+        Ok(authorization)
+    }
+
+    /// Finish the current challenge
+    ///
+    /// This does two things behind the scene
+    /// 1. Notify acme server that challenge is ready and should be verified by the server ([`Self::notify_challenge_ready`])
+    /// 2. Polls the acme server until the server has verified this challenge and has updated its internal status ([`Self::wait_until_challenge_finished`])
+    pub async fn finish_challenge(
+        &self,
+        challenge: &mut server::Challenge,
+    ) -> Result<(), ClientError> {
+        self.notify_challenge_ready(challenge).await?;
+        self.wait_until_challenge_finished(challenge).await
+    }
+
+    /// Notify ACME server that the given challenge is ready
+    ///
+    /// Server will now try to verify this and we should keep polling the server
+    /// with [`Self::wait_until_challenge_finished()`] to wait for the result.
+    pub async fn notify_challenge_ready(
+        &self,
+        challenge: &server::Challenge,
+    ) -> Result<(), ClientError> {
+        let do_request = async || {
+            let response = self.post(&challenge.url, EMPTY_PAYLOAD).await?;
+
+            parse_response::<Empty>(response).await?;
+            Ok(())
+        };
+
+        retry_bad_nonce(do_request).await?;
+
+        Ok(())
+    }
+
+    /// Referesh the given challenge
+    ///
+    /// This returns a duration which the server has requested to wait before calling this again if any
+    pub async fn refresh_challenge(
+        &self,
+        challenge: &mut server::Challenge,
+    ) -> Result<Option<Duration>, ClientError> {
+        let do_request = async || {
+            let response = self
+                .post(&challenge.url, NO_PAYLOAD)
+                .await
+                .context("refresh challenge request")?;
+
+            let retry = get_retry_after_duration(&response);
+            let challenge = parse_response::<server::Challenge>(response).await?;
+            Ok((retry, challenge))
+        };
+
+        let (retry, new) = retry_bad_nonce(do_request).await?;
+        *challenge = new;
+        Ok(retry)
+    }
+
+    /// Poll acme server until the challenge is finished (challenge.status == server::ChallengeStatus::Valid | server::ChallengeStatus::Invalid)
+    pub async fn wait_until_challenge_finished(
+        &self,
+        challenge: &mut server::Challenge,
+    ) -> Result<(), ClientError> {
+        loop {
+            let retry_wait = self.refresh_challenge(challenge).await?;
+
+            match challenge.status {
+                server::ChallengeStatus::Pending | server::ChallengeStatus::Processing => (),
+                server::ChallengeStatus::Valid => return Ok(()),
+                server::ChallengeStatus::Invalid => {
+                    return Err(
+                        OpaqueError::from_display("challenge is detected as invalid").into(),
+                    );
+                }
+            }
+
```

Comment:
seems like a default (the 1s)  that you may want to provide an opt-in way to configure. Btw why the choice for millis(1000) instead of seconds(1)?